### ISSUE TITLE
Improve Trial's --temp-directory option

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3481]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)| @thijstriemstra reported|
|-|-|
|Trac ID|trac#3481|
|Type|enhancement|
|Created|2008-10-21 09:32:37Z|
|Branch|https://github.com/twisted/twisted/tree/trial-tmp-marker-3481|

Trial should drop a marker file into ```_trial_temp``` and refuse to delete and re-create any directory that doesn't have the marker. Docs should updated as well, see [#2589](https://github.com/twisted/twisted/issues/2589) 

Attachments:

* [trial-marker.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/fc9/fc9b86c5fcec82b7df1d0ffbf3e5ebd3cf67fec7/0ed2c95c30d03643ee6860c9bf89107127ed77c4.diff) (2257 bytes) - added by jonathanj on 2009-05-19 17:13:50Z - 
* [trial-marker-1.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/fc9/fc9b86c5fcec82b7df1d0ffbf3e5ebd3cf67fec7/ccc092b0373c3a9e05db89c6a2f23311d279d5ec.diff) (4529 bytes) - added by jonathanj on 2009-06-04 10:26:39Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__3481 3481
type__enhancement enhancement
reporter__thijs thijs
priority__high high
milestone__ 
branch__branches_trial_tmp_marker_3481 branches/trial-tmp-marker-3481
branch_author__jonathanj__thijs jonathanj, thijs
status__closed closed
resolution__fixed fixed
component__trial trial
keywords__ 
time__1224581557000000 1224581557000000
changetime__1247008869000000 1247008869000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|<img alt="moonfallen's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_1"></a>|moonfallen commented|
|-|-|

Yes, let's do this please.  Just had my source directory deleted.  Didn't lose anything important, fortunately.

**Comment by twisted-trac** [comment]
|[<img alt="jonathanj's avatar" src="https://avatars.githubusercontent.com/u/112166?s=50" width="50" height="50">](https://github.com/jonathanj)|@jonathanj removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_3"></a>|@glyph set owner to @jonathanj|
|-|-|

The patch is great!  Could you make a few small adjustments?

  1. In your test case, the name "`my_runner`" violates the coding standard.  It should be "`myRunner`".
  1. In `_removeSafely`'s docstring (thanks for adding that!) you should probably use `L{}` rather than `C{}` to refer to `NoTrialMarker`.
  1. I realize that `trial` is a pile of bad examples as far as this particular request goes, but could you use `twisted.python.filepath.FilePath` rather than `os.path` for path inspection and manipulation?  As a bonus (although this is not required) you could remove the dependence on the `shutil` module, since `FilePath` handles recursive removal itself.
  1. `NoTrialMarker`'s docstring could be more expository.  Especially, why is this exception raised?  When does trial look for a marker?
  1. Also, you might want to follow the example of `_WorkingDirectoryBusy` and make the exception private, since the only method that raises it is also private.  It's always easy to make private stuff public later, less easy to decide it was a bad idea and make it private.

**Comment by twisted-trac** [comment]
|[<img alt="jonathanj's avatar" src="https://avatars.githubusercontent.com/u/112166?s=50" width="50" height="50">](https://github.com/jonathanj)<a name="note_4"></a>|@jonathanj set owner to @glyph|
|-|-|

1.I've changed `_setUpTestdir` and `_removeSafely` to use `FilePath` internally.

Everything else you said made sense, and I've addressed those changes too.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_5"></a>|@thijstriemstra commented|
|-|-|

(In [26970]) Branching to 'trial-tmp-marker-3481'

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_6"></a>|@thijstriemstra commented|
|-|-|

(In [26972]) Update copyright, refs [#3481](https://github.com/twisted/twisted/issues/3481)

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_7"></a>|@thijstriemstra commented|
|-|-|

Buildbot is [happy](http://buildbot.twistedmatrix.com/builders/ubuntu64-py2.5-select/builds/444), looks ready to merge for me.

If this hits the trunk before [[#2589](https://github.com/twisted/twisted/issues/2589)](https://github.com/twisted/twisted/issues/2589) does, then that one should be closed, or marked as fixed with the merge of this ticket. #2589 was created to add a warning to ```trial```'s commandline options, but I don't think we need to add more docs than that.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_8"></a>|@glyph set owner to @thijstriemstra|
|-|-|

Erm... did you review this, or not?  If so, then it should be reassigned to jonathanj, not me.  If not, and you're looking for more feedback from me (although I don't see the question) as the reviewer of a non-committer contribution, you should merge it on behalf of jonathanj :).

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_9"></a>|@glyph commented|
|-|-|

Gah, that comment was badly garbled.  What I meant was:

If you reviewed this, you should be merging it on behalf of jonathanj.

If you just looked at it and gave it a positive comment, but wanted feedback from me, then you should be leaving the review keyword as you reassign it to me.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_10"></a>|@thijstriemstra set status to `closed`|
|-|-|

(In [27098]) Merge trial-tmp-marker-3481: Trial now drops a marker file into _trial_temp and
refuses to delete and re-create any directory that doesn't have the marker.

Author: jonathanj, thijs
Reviewer: glyph
Fixes: [#3481](https://github.com/twisted/twisted/issues/3481)

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
